ENH: Deprecation warnings for `/` integer division when running python -3.

njase opened the pull request:
This feature implements deprecation warning mechanism for integer division in numpy inline with python -3 

See PEP 238 for Python's behaviour, summarised as:
**before 2.2**
Operator //  does not exist
Operator / exists and has mixed behaviour. For integers its floor division, otherwise best possible 

**2.2 onwards until 3.0**
Operator // is now defined for floor division irrespective of numerical data type
Operator /  - same as earlier 
Using -3 switch will report warning if mismatch with future (3.x) behaviour

**Starting from 3.0**
Operator // for floor division, same as earlier
/ best possible division

charris commented on numpy/core/src/umath/ufunc_type_resolution.c:
```diff
@@ -1210,6 +1211,43 @@ type_reso_error: {
+
+    /* If both types are integer, warn the user, same as python does */ 
+    if ( Py_DivisionWarningFlag &&
+         ( type_num1 == NPY_BOOL || type_num1 == NPY_BYTE || type_num1 == NPY_UBYTE || \
```
All lines < 80 characters.

charris replied:
Maybe make a list/dict of types and use `type_num1 in ...`

njase replied:
Accepted, done

charris commented on numpy/core/src/umath/ufunc_type_resolution.h:
```diff
@@ -65,6 +65,13 @@ PyUFunc_MultiplicationTypeResolver(PyUFuncObject *ufunc,
 
 NPY_NO_EXPORT int
+PyUFunc_MixedDivisionTypeResolver(PyUFuncObject *ufunc,
+                                NPY_CASTING casting,
```
Indentation is off, should line up under `PyUFuncObject`.

njase replied:
Accepted, corrected in the whole file

seberg commented:
Maybe a start:
```diff
diff --git a/numpy/testing/nosetester.py b/numpy/testing/nosetester.py
index c30c0ee..87bc344 100644
--- a/numpy/testing/nosetester.py
+++ b/numpy/testing/nosetester.py
@@ -412,6 +412,17 @@ class NoseTester(object):
                 from ..distutils import cpuinfo
             sup.filter(category=UserWarning, module=cpuinfo)
 
+            # Filter out deprecation warnings due to the -3 flag to python 2.
+            if sys.version_info.major == 2 and sys.py3kwarning:
+                # This is very specific, so using the fragile module filter
+                # is fine.
+                import threading
+                sup.filter(DeprecationWarning,
+                           r"sys\.exc_clear\(\) not supported in 3\.x",
+                           module=threading)
+                sup.filter(DeprecationWarning, message="in 3\.x, __setslice__")
+                sup.filter(DeprecationWarning, message="in 3\.x, __getslice__")
+
             # Filter out some deprecation warnings inside nose 1.3.7 when run
             # on python 3.5b2. See
             #     https://github.com/nose-devs/nose/issues/929
```
If the slice things comes from a single module, could use a module filter. But my guess is it probably does not.

seberg commented on numpy/testing/nosetester.py:
```diff
@@ -412,6 +412,22 @@ def test(self, label='fast', verbose=1, extra_argv=None,
+                sup.filter(DeprecationWarning, message="buffer\(\) not supported in 3\.x")
+                sup.filter(DeprecationWarning, message="CObject type is not supported in 3\.x")
+                
+   
```
Two many empty lines.

njase replied:
Done

seberg commented on numpy/core/tests/test_regression.py:
```diff
@@ -2115,6 +2115,10 @@ def __array__(self,*args,**kwargs):
             if sys.version_info[0] >= 3:
                 assert_raises(TypeError, f, lhs, rhs)
+            elif sys.version_info[0] < 3 and sys.py3kwarning:
+                warnings.filterwarnings('error',category=DeprecationWarning)
```
please put the spaces here as per PEP8 (just nitpicking, but it seemed like a few places, not just here).

njase replied:
Sorry but could you point out to some locations? I could not spot violations (4 spacing rules, max line width, etc are followed)

seberg replied:
There should be a space after the comma.

seberg commented on numpy/core/tests/test_regression.py:
```diff
@@ -2115,6 +2115,10 @@ def __array__(self,*args,**kwargs):
+            elif sys.version_info[0] < 3 and sys.py3kwarning:
+                warnings.filterwarnings('error',category=DeprecationWarning)
+                assert_raises(DeprecationWarning, f, lhs, rhs)
+                warnings.filterwarnings('ignore',category=DeprecationWarning)
```
The test is failing due to this. Don't use "ignore" filters, they are horribly broken in versions before python 3.4.x if you want to reliably test them. Also filters should always be inside a context (either catch_warnings, or our own suppress_warnings context).

My guess is: put this test into a warnings context. Then add a new test if you want to check whether it actually gives the warning when py3kwarning is active.

njase replied:
Hmm, I see. I also wondered why so many tests in a single file are unexpectedly failing.

On second thought, should I just ignore this test during -3 switch? 
The test will anyhow run without -3 as part of the whole test suite..?

All other comments I'll take care asap

seberg replied:
Since the tests run anyhow, I am OK with skipping when its easier. Though suppression is cleaner and seems preferable to me if it is straight forward.

seberg commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -6457,7 +6457,7 @@ class TestArrayPriority(TestCase):
         ]
 
-    if sys.version_info[0] < 3:
+    if sys.version_info[0] < 3 and not sys.py3kwarning:
```
Could maybe use a comment. Or put the failing tests into a suppress_warnings context. I think I am fine with just doing it like this, but an explanation might help in the long run.

seberg reviewed (commented):
Only read the tests stuff, seems like nothing odd came up within numpy (not that I expected it), so mostly nitpicks.

seberg commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -6457,6 +6457,8 @@ class TestArrayPriority(TestCase):
         ]
 
+    #See #7949. Dont use "/" operator With -3 switch, since python reports it
+    #as a DeprecationWarning
```
space after the `#` in PEP8

njase commented:
All comments taken care. Checks passed. Hope the changes can be accepted now. 
Thanks @seberg

seberg commented:
You are too fast :). I still have to read all, so no need to do all yet, but should still add a release note and squash the commits. Thanks for bwing patient with the nitpicks though ;).

njase commented:
I'll squash the commits when there are no more comments, and you give a go ahead.
For release notes: how to decide which version to update?https://github.com/numpy/numpy/tree/master/doc/release

seberg commented on numpy/core/tests/test_deprecations.py:
```diff
@@ -646,6 +646,23 @@ def foo():
+            import operator as op
+            for dt in deprecated_types:
+                a = np.array([1,2,3], dtype=dt)    
+                self.assert_deprecated(op.div, args=(a,a)) 
```
space after comma :)

njase replied:
Same as earlier.

seberg commented on numpy/core/tests/test_deprecations.py:
```diff
@@ -646,6 +646,23 @@ def foo():
+        if sys.version_info[0] < 3 and sys.py3kwarning:
+            import operator as op
+            for dt in deprecated_types:
+                a = np.array([1,2,3], dtype=dt)    
```
Even here ;) space after comma. Seriously, though, if you don't want to fix, its OK.

njase replied:
PEP8 discourages extraneous whitespace immediately before comma,semicolon (see second bullet in section Pet Peeves)
In this case, space is after comma which should be acceptable since it avoids cluttering (IMHO)

Please correct me if I am wrong. For consistency and code quality, I am ready to change my coding habits :)

seberg replied:
Don't worry about it, I meant within the list, which sometimes is really OK to not put the space after comma. Not sure, I can imagine PEP8 says no space after the comma is OK in some cases (e.g. I also sometimes tend to skip it in indexing operations).

seberg commented on numpy/core/tests/test_regression.py:
```diff
@@ -2115,7 +2115,9 @@ def __array__(self,*args,**kwargs):
-            else:
+            elif not sys.py3kwarning:
+                # With -3 switch in python 2, DeprecationWarning is raised
+                # which we are not interested in
```
A bit confused what warning this is, but OK.

njase replied:
I could not add a warning context here as you commented last time. So the idea was to skip this test under -3 switch.

seberg commented on numpy/testing/nosetester.py:
```diff
@@ -411,7 +411,20 @@ def test(self, label='fast', verbose=1, extra_argv=None,
+                sup.filter(DeprecationWarning, message="in 3\.x, __getslice__")
+                sup.filter(DeprecationWarning, message="buffer\(\) not supported in 3\.x")
+                sup.filter(DeprecationWarning, message="CObject type is not supported in 3\.x")
+                sup.filter(DeprecationWarning, message="comparing unequal types not supported in 3\.x")
```
I might still have a look what the last two things filter later, but I trust you its all good.

njase replied:
With the existing unit tests (before this change), overall 6 types of errors were reported with -3 switch. Those reported as "classic int division" have been fixed.
- "CObject.." API are deprecated, so I ignored such errors.
- "unequal types..". I could not figure out how these cases pass with actual python 3 version. But they do, so I ignored them with -3 switch

seberg commented:
The next release is 1.12, so put it into the 1.12.0 release notes.

njase commented on numpy/core/src/umath/ufunc_type_resolution.c:
```diff
@@ -1210,6 +1211,67 @@ type_reso_error: {
     }
 }
 
+int is_integer_type(int type_num)
```
That was nice :)

njase commented on numpy/core/src/umath/ufunc_type_resolution.c:
```diff
@@ -1210,6 +1211,67 @@ type_reso_error: {
+    /* If both types are integer, warn the user, same as python does */ 
+    if ( Py_DivisionWarningFlag &&
+         (1 == is_integer_type(type_num1)) &&
+         (1 == is_integer_type(type_num2)) )
```
accepted

seberg commented:
Your probably already got it, but you will have to rebase your changes on master to avoid conflicts in the release notes (unfortunatly, those happen quite quickely). I also think you may have check for the BOOL typenum as well (there is also a macro for that).

njase commented:
Thanks @seberg.
Changes rebased with master
BOOL check added (I learnt it the hard way!) - thanks for pointing out
Should I squash my changes now?

seberg commented:
Thanks  @njase, I think I am done with nitpicking. If you squash it I will merge it (unless someone else has an opinion about it).

charris commented:
Needs commit squashing, do `git rebase -i HEAD~12` and follow directions. After squashing do a force push `git push -f origin HEAD`.

homu commented:
:umbrella: The latest upstream changes (presumably 0a02bb6) made this pull request unmergeable. Please resolve the merge conflicts.

njase commented:
sorry for delay, was quite occupied for last few weeks.
@charris, how do I squash now since my commits are mingled with other's commits when I've merged upstream at least 2 times?

..also, git hub help says for the admin (https://github.com/blog/2141-squash-your-commits"
"Only allow squash commits : This is a new option which lets you force commit squashing on all pull requests merged via the merge button."
Is such an option visible to you for my branch?

stuarteberg commented:
Hi @njase, on my machine, I was able to rebase your branch on the latest master and also squash all the commits into one.

For reference, here's a log of the commands I used.

I started by making sure that my local `numpy` repo was in sync with the latest version on github:

```
git p
```

I started by fetching your commits so I could check out your branch:

```bash
git remote add saurabh ssh://git@github.com/njase/numpy
git fetch saurabh
git checkout enh_7949
```

Then I rebased your branch on the latest version
```
git rebase master
# (Here, I had to fix one tiny merge conflict during the rebase operation.)
git rebase --continue
```

stuarteberg commented:
Hi @njase,

>how do I squash now since my commits are mingled with other's commits

On my machine, I was able to rebase your branch on the latest master and also squash all the commits into one.  (See appendix, below.)

I pushed the results to a branch on my fork:
https://github.com/stuarteberg/numpy/tree/enh_7949_squashed

If that looks okay to you, you can update this PR as follows:

```bash
git checkout enh_7949
git branch backup_enh7949 # Save a copy of your branch, just in case
git remote add stuart https://github.com/stuarteberg/numpy
git reset --hard stuart/enh_7949_squashed
git push -f https://github.com/njase/numpy enh_7949
```

----

Appendix
---------


For reference, here's a log of the commands I used to rebase/squash your branch.

I started by making sure that my local `numpy` repo was in sync with the latest version on github:

```bash
cd numpy
git checkout master
git pull origin master
```

Then I fetched and checked out your branch:

```bash
git remote add saurabh ssh://git@github.com/njase/numpy
git fetch saurabh
git checkout enh_7949
```

Then I rebased your branch on the latest commit from master:

```bash
git rebase master
# (Here, I had to fix one tiny merge conflict during the rebase operation.)
git rebase --continue
```

At this point, the git history looks as if all of your commits were just now added, linearly, on top of today's master branch.

To squash them into a single commit, I used `git rebase` again, this time with `--interactive` mode, which allows you to select commits to combine:


```
git rebase --interactive master
```

njase committed:
ENH: Deprecation warnings for `/` integer division when running python -3

When python is invoked with switch -3, it emits waring "classic int division"
for strict integer divisions. The same behavior is now implemented to numpy
with this fix

njase commented:
Thanks @stuarteberg
Its all done from my side. Ready for inclusion into main release

seberg commented:
OK, lets give it a shot. Thanks @njase!

njase commented:
Thanks. I enjoyed working on this and look forward to contribute more in future.
